Sources/UnleashProxyClientSwift/Metrics.swift

```diff
@@ -0,0 +1,151 @@
+import Foundation
+import SwiftEventBus
+
+extension Date {
+    func iso8601String() -> String {
+        let formatter = ISO8601DateFormatter()
+        return formatter.string(from: self)
+    }
+}
+
+struct ToggleMetrics {
+    var yes: Int = 0
+    var no: Int = 0
+    var variants: [String: Int] = [:]
+
+    func toJson() -> [String: Any] {
+        ["yes": yes, "no": no, "variants": variants]
+    }
+}
+
+struct Bucket {
+    private let clock: () -> Date
+    var start: Date
+    var stop: Date?
+    var toggles: [String: ToggleMetrics] = [:]
+
+    init(clock: @escaping () -> Date) {
+        self.clock = clock
+        start = clock()
+    }
+
+    mutating func closeBucket() {
+        stop = clock()
+    }
+
+    func isEmpty() -> Bool {
+        toggles.isEmpty
+    }
+
+    func toJson() -> [String: Any] {
+        let mappedToggles = toggles.mapValues { $0.toJson() }
+        return [
+            "start": start.iso8601String(),
+            "stop": stop?.iso8601String() ?? "",
```

Comment:
Are there any cases where we wouldn't have a `stop` when we serialize?

Reply:
we always stop the bucket before sending. With the current type-level model it's impossible to express it that's why we have this weird condition as in other APIs. I didn't want to introduce a separate type for StoppedBucket as it would deviate from how we do other APIs